pkg/container-hook/tracer.go

```diff
@@ -537,44 +555,41 @@ func (n *ContainerNotifier) monitorRuntimeInstance(bundleDir string, pidFile str
 		// podman might install config.json in the userdata directory
 		configJSONPath = filepath.Join(bundleDir, "userdata", "config.json")
 		if _, err := os.Stat(configJSONPath); errors.Is(err, os.ErrNotExist) {
-			pidFileDirNotify.File.Close()
 			return fmt.Errorf("config not found at %s", configJSONPath)
 		}
 	}
-	err = pidFileDirNotify.Mark(unix.FAN_MARK_ADD|unix.FAN_MARK_IGNORED_MASK, unix.FAN_ACCESS_PERM, unix.AT_FDCWD, configJSONPath)
+	err = n.pidFileDirNotify.Mark(unix.FAN_MARK_ADD|unix.FAN_MARK_IGNORED_MASK, unix.FAN_ACCESS_PERM, unix.AT_FDCWD, configJSONPath)
 	if err != nil {
-		pidFileDirNotify.File.Close()
 		return fmt.Errorf("marking %s: %w", configJSONPath, err)
 	}
 
-	// similar to config.json, we ignore passwd file if it exists
-	passwdPath := filepath.Join(bundleDir, "passwd")
-	if _, err := os.Stat(passwdPath); !errors.Is(err, os.ErrNotExist) {
-		err = pidFileDirNotify.Mark(unix.FAN_MARK_ADD|unix.FAN_MARK_IGNORED_MASK, unix.FAN_ACCESS_PERM, unix.AT_FDCWD, passwdPath)
+	// To reduce noise, ignore additional files (best effort)
+	ignoreFileList := []string{
+		"passwd",
+		"log.json",
+		"runtime",
+	}
+	for _, ignoreFile := range ignoreFileList {
+		ignoreFilePath := filepath.Join(bundleDir, ignoreFile)
+		err := n.pidFileDirNotify.Mark(unix.FAN_MARK_ADD|unix.FAN_MARK_IGNORED_MASK, unix.FAN_ACCESS_PERM, unix.AT_FDCWD, ignoreFilePath)
 		if err != nil {
-			pidFileDirNotify.File.Close()
-			return fmt.Errorf("marking passwd path: %w", err)
+			log.Debugf("marking %s: %v", ignoreFilePath, err)
 		}
 	}
 
-	n.wg.Add(1)
-	go func() {
-		defer n.wg.Done()
-		defer pidFileDirNotify.File.Close()
-		for {
-			stop, err := n.watchPidFileIterate(pidFileDirNotify, bundleDir, configJSONPath, pidFile, pidFileDir)
-			if n.closed.Load() {
-				return
-			}
-			if err != nil {
-				log.Warnf("error watching pid: %v\n", err)
-				return
-			}
-			if stop {
-				return
-			}
-		}
-	}()
+	// cri-o appends userdata to bundleDir,
+	// so we trim it here to get the correct containerID
+	containerID := filepath.Base(filepath.Clean(strings.TrimSuffix(bundleDir, "userdata")))
+
+	n.pendingMu.Lock()
+	defer n.pendingMu.Unlock()
+	n.pendingContainers[pidFile] = &pendingContainer{
+		id:             containerID,
+		bundleDir:      bundleDir,
+		configJSONPath: configJSONPath,
+		pidFile:        pidFile,
+		pidFileDir:     pidFileDir,
+	}
```

Comment:
failed containers will never be removed from this map, do you think it could be an issue? Extra memory usage?

Do you think we should implement some clean up logic (can be done later on if you think it's worth it)

Reply:
Fixed in the last commit